Debian updates

whitty opened the pull request:
Hi - me again - (sorry).

I believe this addresses issues #43 and #120 (by switching to "native build format") - although I'm concerned that someone _was_ able to generate a ppa at some point so perhaps there is some source build method I don't know that _does_ work.

_This_ method is compatible with ppa (source) autobuilding (aka launchpad automatic nightlies).  It also works for the instructions in the readme.

So if someone knows better than me on how this should be formed I'm open to being corrected.

Summary:

* Adds the newer modules to dkms
* Switches dkms versioning to come directly from debian scripts (no need to update separately)
* Adds rtw89-firmware package
* conflicts with the other rlwfinger/tw8852be project (ie force uninstalls it)
* defines modaliases (theoretically ubuntu could auto install the package on detecting hardware)
* restricts build version for kernels 5.7+

If you don't want the changes, I'll maintain them in launchpad to build the ppas.

whitty commented on debian/rtw89-dkms.dkms:
```diff
@@ -1,14 +1,23 @@
+DEST_MODULE_LOCATION[7]="/updates/"
 AUTOINSTALL="YES"
+# Set the acceptable kernel values - 5.7+
+BUILD_EXCLUSIVE_KERNEL="^(5\.0[7-9]+\.|5\.[1-9][0-9]+\.|6.[0-9]+\.)"
```
OK - I see the error here - will re-push

whitty replied:
I tested this against ubuntu focal (5.4) - but I didn't have a 5.8+ kernel around - the old one would exclude all 5.[0-9] versions, this one should allow 5.[7-9]

mbanck commented:
I think those changes are going in the right direction. I packaged rwt89 for personal use a while ago, and had some of them as well, but this looks much cleaner.

> I believe this addresses issues #43 and #120 (by switching to "native build format") - although I'm concerned that someone _was_ able to generate a ppa at some point so perhaps there is some source build method I don't know that _does_ work.

One underlying issue seems to be that the version in `debian/changelog` has no correspondence with upstream, it seems like it fell out of a packaging template (1.0.0).

The usual practice would be to name the upstream version according to some upstream tag (there hasn't been one in a while), and then possibly add a git snapshot ref/date, like `v7-20211023+git2022114` or so.

The fact that the Debian packaging isn't being updated regularly hints that this should maybe not be a native package after all, or what is the plan going forward?

>     * Adds rtw89-firmware package

What is the `rtw89-firmware` `pre/postinst` `dpkg-divert` stuff for? Is that standard practice for firmware packages?
 
>     * conflicts with the other rlwfinger/tw8852be project (ie force uninstalls it)

What is this?

lwfinger commented:
I have merged these changes. As I have never provided the stuff in the debian directory, but only merged what I have been given, I am not surprised that it has gotten bitrotted.

You can argue the details among yourselves.

whitty commented:
> I think those changes are going in the right direction. I packaged rwt89 for personal use a while ago, and had some of them as well, but this looks much cleaner.
> 
> > I believe this addresses issues #43 and #120 (by switching to "native build format") - although I'm concerned that someone _was_ able to generate a ppa at some point so perhaps there is some source build method I don't know that _does_ work.
> 
> One underlying issue seems to be that the version in `debian/changelog` has no correspondence with upstream, it seems like it fell out of a packaging template (1.0.0).
> 
> The usual practice would be to name the upstream version according to some upstream tag (there hasn't been one in a while), and then possibly add a git snapshot ref/date, like `v7-20211023+git2022114` or so.

I do want to look into this - if I can understand the current module versioning.

When you say "upstream" what do you mean exactly?  I'm treating this as the definitive source for the `-dkms` package (there's separate definitive source in the kernel -  but presumably you could use that without this package).  I believe some kernel modules use kernel-module versioning and that can probably be referenced by dkms.  I don't think this module does though.

When I publish this here: https://launchpad.net/~greg-whiteley/+archive/ubuntu/rtw89/ - I generate a date-based version, but its based off the fake debian version.

(now that this has merged I can update this to auto-build - rather than requiring manual work)

> The fact that the Debian packaging isn't being updated regularly hints that this should maybe not be a native package after all, or what is the plan going forward?

As far as I can work out "native" in this sense - means the only source for the deb is this source control (ie this github repo).  The alternative is to use a source "orig tarball" which needs a version number.  There are tools to "fake" this, but "native" is simpler.

> > ```
> > * Adds rtw89-firmware package
> > ```
> 
> What is the `rtw89-firmware` `pre/postinst` `dpkg-divert` stuff for? 

This does two things: - 

1. when `rtw89-firmware` is installed it will move the existing named firmware out of the way
2. and symlink in its place to the firmware from the `rtw89-firmware` package

When the `linux-firmware` package is installed/updated it will

1. remove the diversion to  the `rtw89-firmware` package
2. install the new firmware package
3. repeat the steps above

> Is that standard practice for firmware packages?

I couldn't find a standard practice - there's an atomic `linux-firmware` package and I don't want to "conflict" with it, but I want to be able to update any firmwares that might also be present.  In my case (22.04) there were "a" firmwares available.

I looked around to see if there was an "alternate" directory (like the `updates` directory in `/lib/modules` but couldn't find anything).  For the most part `/lib/firmware` paths are baked into the kernel.

If someone knows a better way, please let me know.

> > ```
> > * conflicts with the other lwfinger/rtw8852be project (ie force uninstalls it)
> > ```
> 
> What is this?

Fixed the typo to "lwfinger/rtw8852be" - an existing github repo that had rtw8852be WiFi support (my card) before this repo did.  It is now considered obsolete.

lwfinger commented:
When I say 'upstream', I mean the content of the wireless-next repo. It feeds into net-next => linux-next, which contains the patches for the next kernel.

whitty commented:
Let me look through - but first blush suggests such a version will have to be manually patched into the deb changelog (which I understand @lwfinger isn't particularly interested in maintaining).

Perhaps I can write a script to make it more palatable - point it at a wireless-next repo and it inserts a "bumped upstream version' changelog entry.  `git describe` today returns `wireless-next-2022-10-28-10582-ga1d0b9535cd8` I could try to script a changelog update using that to produce a version (probably sans `wireless-next-` ).

The alternative is to manage tagging and versioning directly in this repo
